### PR TITLE
CASTOR reconstruction and simulation updates

### PR DESCRIPTION
**Pull request by hvanhaev**
This pull request contains updates for CASTOR reconstruction and simulation.
After integration of this in master, the request is that these changes are also allowed to be back ported to the 76X and 80X branches for use with 2015 and 2016 data.

Reconstruction:
* Fix of RecHitCorrector class: 
   - this class is not used during default reconstruction sequence, but it is important for analyses, as it can very easily reprocess CASTOR objects on the fly during ntuple steps. 
   - the code is updated with: memory leak fix, removal of obsolete factor 2, cleaning up code following comments from a previous PR (https://github.com/cms-sw/cmssw/pull/21860#issuecomment-357963218). 
   - I tried to replace the for loops with range-based for loop, but could not figure it out for the line "for (unsigned int i=0;i<rechits->size();i++)". If someone can provide an alternative for this then please let me know.

Simulation:
* A long time ago we discovered a small bug in the CASTOR noise addition (during digi step), which would overestimate the noise, especially in case of pA and AA collisions. Now finally I managed to prepare the code for submission. This change has already been validated within CASTOR group. These classes are used during production whenever the CASTOR simulation is executed/included in the config.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by hvanhaev** [comment]
Ok, so I unintentionally added a commit to this pull request on a different topic, didn't know GitHub did this automatically. Anyway, it also needs to find its way into CMSSW, but it is related to simulation group: a long time ago we discovered a small bug in the CASTOR noise addition (during digi step), which would overestimate the noise, especially in case of pA and AA collisions. Now finally I managed to prepare the code for submission. This change has already been validated within CASTOR group. Also here, we eventually like to be able to back port this into 76X and 80X branches. Please let me know if I need to separate the topics(commits) or not. Apologies if I caused any confusion.

**Comment by perrotta** [comment]
github don't do it automatically, but only if you push the change into the branch ;-)

In principle, approval and integration can proceed faster if different and unrelated issues are kept separated, and this is what I'd suggest here.

Anyhow, if both reco and sim updates can be considered complete here (modulo the review) perhaps you can also stick on this single pull request. If you do so, however, please update the title and add a complete description of the two by editing the description window on top of this thread

By the way, could you please also specify in the description what is the expected usage for the RecHitCorrector class and all other classes in ```RecoLocalCalo/Castor/src```? 
If they are supposed to be used in production, please convert them into thread safe stream modules (as in was already suggested in a previous incarnation of this PR in https://github.com/cms-sw/cmssw/pull/21860#issuecomment-357963218); and moreover, please move the class declarations in the ```\interface``` subdirectory, and only keep the implementation in ```\src``` (this suggestion holds also if these classes are not meant to be used in any central production)

**Comment by hvanhaev** [comment]
Hi,

I changed the title and updated the description at the top.
Regarding the rest of RecoLocalCalo/Castor: I wanted to avoid to change everything, if I need to clean up the other files it will take me more time. In particular CastorTowerProducer is important yes, since it is executed during default production. The others are not. Concerning your previous comments, I implemented those on the RecHitCorrector, as the PR was about that. Not sure what I still have to convert in the other files to be thread safe there, after quickly looking it only involves removing commented lines it seems.

I will have a look at the other files in /src, once ready I will include it here or in a new PR.

On 8 Mar 2018, at 11:46, perrotta <notifications@github.com<mailto:notifications@github.com>> wrote:


github don't do it automatically, but only if you push the change into the branch ;-)

In principle, approval and integration can proceed faster if different and unrelated issues are kept separated, and this is what I'd suggest here.

Anyhow, if both reco and sim updates can be considered complete here (modulo the review) perhaps you can also stick on this single pull request. If you do so, however, please update the title and add a complete description of the two by editing the description window on top of this thread

By the way, could you please also specify in the description what is the expected usage for the RecHitCorrector class and all other classes in RecoLocalCalo/Castor/src?
If they are supposed to be used in production, please convert them into thread safe stream modules (as in was already suggested in a previous incarnation of this PR in #21860 (comment)<https://github.com/cms-sw/cmssw/pull/21860#issuecomment-357963218>); and moreover, please move the class declarations in the \interface subdirectory, and only keep the implementation in \src (this suggestion holds also if these classes are not meant to be used in any central production)

**Comment by perrotta** [comment]
> I changed the title and updated the description at the top.

Thank you

> Regarding the rest of RecoLocalCalo/Castor: I wanted to avoid to change everything, if I need to clean
> up the other files it will take me more time. In particular CastorTowerProducer is important yes, since it
> is executed during default production. The others are not. Concerning your previous comments, I
> implemented those on the RecHitCorrector, as the PR was about that. Not sure what I still have to
> convert in the other files to be thread safe there, after quickly looking it only involves removing
>  commented lines it seems.

Ok, fine: CastorTowerProducer is already a stream module. I agree with you about not touching now the other modules. Please also disregard my previous suggestion of moving the class declaration into header files in a ```\interface``` subdirectory: since they are producers they don't need to be exposed, and it is correct that they don't have an interface that can be included.

However, since you are touching here RecHitCorrector, please make it a stream module also: it is true that it is not used now in production, but since it stay in a \src area let it move away from being a legacy module (otherwise: move it into a \test area)

File: RecoLocalCalo/Castor/src/RecHitCorrector.cc [hunk 1]
```diff
@@ -114,59 +114,47 @@ RecHitCorrector::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
...
    
    for (unsigned int i=0;i<rechits->size();i++) {
    	CastorRecHit rechit = (*rechits)[i];
-	//std::cout << "rechit energy = " << rechit.energy() << std::endl;
 	double fC = factor_*rechit.energy();
 	double time = rechit.time();
-	//std::cout << "rechit energy(fC) = " << fC << " time = " << time << std::endl;
 	
 	// do proper gain calibration reading the latest entries in the condDB
 	const CastorCalibrations& calibrations=conditions->getCastorCalibrations(rechit.id());
```
Comment (perrotta) [inline RecoLocalCalo/Castor/src/RecHitCorrector.cc]: ```calibrations``` and ```capid``` below are only needed ```if (doInterCalib_)```
You can move their definition under that condition, e.g.:
```
      double time = rechit.time();
      double correctedenergy = factor_*rechit.energy();
      if (doInterCalib_) {
	 // do proper gain calibration reading the latest entries in the condDB
	 int capid = 0; // take some capid, gains are the same for all capid's
	 const CastorCalibrations& calibrations=conditions->getCastorCalibrations(rechit->id());
         correctedenergy *= calibrations.gain(capid);
      }

```

File: RecoLocalCalo/Castor/src/RecHitCorrector.cc [hunk 2]
```diff
@@ -114,59 +114,47 @@ RecHitCorrector::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
...
 	std::vector<DetId> channels = myqual->getAllChannels();
-	//std::cout << "number of specified quality flags = " << channels.size() << std::endl;
-	for (std::vector<DetId>::iterator channel = channels.begin();channel !=  channels.end();channel++) {	
-		if (channel->rawId() == detcell.rawId()) {
-			const CastorChannelStatus* mydigistatus=myqual->getValues(*channel);
-			//std::cout << "CastorChannelStatus = " << mydigistatus->getValue() << std::endl;
+	for (auto channel : channels) {	
+		if (channel.rawId() == detcell.rawId()) {
+			const CastorChannelStatus* mydigistatus=myqual->getValues(channel);
 			if (mydigistatus->getValue() == 2989) ok = false; // 2989 = BAD
```
Comment (perrotta) [inline RecoLocalCalo/Castor/src/RecHitCorrector.cc]: You could even ```break``` here: no need to go through all remaining channels if ```ok``` is already ```false```

**Comment by perrotta** [comment]
ping @hvanhaev

**Comment by hvanhaev** [comment]
Thanks, for the reminder, I hope I can implement the suggestions in course of the next days...

On 15 Mar 2018, at 11:54, perrotta <notifications@github.com<mailto:notifications@github.com>> wrote:


ping @hvanhaev<https://github.com/hvanhaev>

**Comment by fabiocos** [comment]
@hvanhaev I understand that this PR is targeting studies for past years analysis, and it will have to be considered for 10_2_X at least. Could you please confirm or clarify?

**Comment by hvanhaev** [comment]
Dear Fabio,

Yes indeed, if possible & allowed, it would be good to have:
- it implemented in a next CMSSW_10_X_Y so we can use it when CASTOR will participate in 2018 Heavy Ion running.
- it back ported to a release to be used with 2015 data (I assume 76X) and 2016 Heavy Ion data (I assume 80X).

This PR does not contain new features, but does have fixes for bugs that are important for CASTOR. It should also only affect CASTOR simulation and reconstruction (so no interference with other subsystems).

Cheers,
Hans


On 16 Mar 2018, at 14:14, Fabio Cossutti <notifications@github.com<mailto:notifications@github.com>> wrote:


@hvanhaev<https://github.com/hvanhaev> I understand that this PR is targeting studies for past years analysis, and it will have to be considered for 10_2_X at least. Could you please confirm or clarify?

**Comment by perrotta** [comment]
Hans Van Haevermaet <notifications@github.com> ha scritto:

> Dear Fabio,
>
> Yes indeed, if possible & allowed, it would be good to have:
> - it implemented in a next CMSSW_10_X_Y so we can use it when CASTOR  
> will participate in 2018 Heavy Ion running.

As far I know the discussion whether to install CASTOR during TS2 is  
still ongoing: or was a decision taken by now?

By the way, all the CASTOR code that will eventually be allowed in the  
production sequences (not only these classes, I imagine) must be  
updated along the lines that we started to discuss in this PR

> - it back ported to a release to be used with 2015 data (I assume  
> 76X) and 2016 Heavy Ion data (I assume 80X).
>
> This PR does not contain new features, but does have fixes for bugs  
> that are important for CASTOR. It should also only affect CASTOR  
> simulation and reconstruction (so no interference with other  
> subsystems).
>
> Cheers,
> Hans
>
>
> On 16 Mar 2018, at 14:14, Fabio Cossutti

**Comment by hvanhaev** [comment]
As far I know the discussion whether to install CASTOR during TS2 is
still ongoing: or was a decision taken by now?

I don’t think it is indeed decided whether it will be installed during TS2 or TS3, but I do know that the aim of the CASTOR group is to participate in the HIN run this year, which is after TS3.
In any case, your comment on the previous PR implied that I first need to submit to master, and then request back ports...

By the way, all the CASTOR code that will eventually be allowed in the
production sequences (not only these classes, I imagine) must be
updated along the lines that we started to discuss in this PR

Basically all CASTOR code (except this RecHitCorrector, InvalidDataFilter and old cluster code) is already in production sequences since many years.
If I understand you correctly you like that we do a complete cross check of all our code (sim, digi, reconstruction, alca, …) to update according to new C++ standards?



On 16 Mar 2018, at 15:01, perrotta <notifications@github.com<mailto:notifications@github.com>> wrote:


Hans Van Haevermaet <notifications@github.com<mailto:notifications@github.com>> ha scritto:

> Dear Fabio,
>
> Yes indeed, if possible & allowed, it would be good to have:
> - it implemented in a next CMSSW_10_X_Y so we can use it when CASTOR
> will participate in 2018 Heavy Ion running.

As far I know the discussion whether to install CASTOR during TS2 is
still ongoing: or was a decision taken by now?

By the way, all the CASTOR code that will eventually be allowed in the
production sequences (not only these classes, I imagine) must be
updated along the lines that we started to discuss in this PR

> - it back ported to a release to be used with 2015 data (I assume
> 76X) and 2016 Heavy Ion data (I assume 80X).
>
> This PR does not contain new features, but does have fixes for bugs
> that are important for CASTOR. It should also only affect CASTOR
> simulation and reconstruction (so no interference with other
> subsystems).
>
> Cheers,
> Hans
>
>
> On 16 Mar 2018, at 14:14, Fabio Cossutti

**Comment by hvanhaev** [comment]
Hi Andrea, 
I committed your proposed changes to RecHitCorrector. Now it should be a stream module (?), I also removed the unused begin/endJob functions. Please let me know if you have further comments.
Cheers,
Hans

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
+1
- Changes/fixes inside ```RecoLocalCalo/Castor/src/RecHitCorrector.cc``` are fine and comply with the requests received; that package does not enter currently in the standard reco sequence
- Changes are observed in the CastorRecHits and CastorTower reco outputs: they are due to the fixes applied here to the CASTOR Simulation packages, e..g. in TTbar13TeVPU2017 wf 10224.0:
image

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
@fabiocos : I think this can be merged in the master, if you don't see issues for it

**Comment by fabiocos** [comment]
please test

**Comment by fabiocos** [comment]
@civanch @ianna CASTOR is presently not installed, it will be for the HI run, but it was kept anyway in the 2018 geometry configuration, even if TOTEM T1 and T2 were removed (see https://indico.cern.ch/event/714398/contributions/2938076/attachments/1618558/2573716/VI20180316sum.pdf ). I understand that this was done under the assumption that there is not effect to be expected on HF from possible back-scattering from it, right? So it is not useful but it does not harm either, is this the assumption?

@hvanhaev @slava77 @perrotta I see that also the 2017 workflows are affected by these changes, no era was used. From the initial comment I understand that the updates should be considered as bug fixes to be back-ported in 2015 and 2016 simulations, so it is ok that not only the 2018 behaviour is modified. Do you confirm?

**Comment by civanch** [comment]
@fabiocos , both CASTOR and ZDS simulations are performed using shower libraries. This means that if a particle (with whatever energy) hit one of this calorimeters it is killed and set of hits are produced. So, they physically cannot backscatter to HF. In the #22145 this will not be true for Hyperons - they should first decay and only after SL will be activated, however, it is difficult to assume that a Hyperon will backscatter and path few meters instead of decay.

Concerning TOTEM: last year we missed the fact, that T1 detector was removed before 2017 run. In 2018 T2 was removed. The most effect come from T1 removal, because it is in front. Note, that HF shower library was created in a special simulation when primary particles hit HF alone, so HF SL is more close to the reality, when TOTEM T1 and T2 are removed from the geometry.

**Comment by perrotta** [comment]
@fabiocos : yes, the changes in 2017 wfs (also observed in https://github.com/cms-sw/cmssw/pull/22424#issuecomment-374907797) come from the fixes for the previously overestimated noise in the CASTOR simulation, and they are supposed to be applied whenever the CASTOR simulation is run

**Comment by fabiocos** [comment]
@hvanhaev @civanch @perrotta thanks for the clarifications, I will integrate this PR with other SIM-related developments after 10_2_0_pre1 is built

**Comment by fabiocos** [comment]
+1

**Pull request merged**